clippy_lints/src/methods/mod.rs

```diff
@@ -2142,6 +2164,48 @@ fn lint_filter_map_flat_map<'a, 'tcx>(
     }
 }
 
+/// lint use of `flat_map` for `Iterators` where `flatten` would be sufficient
+fn lint_flat_map<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx hir::Expr, flat_map_args: &'tcx [hir::Expr]) {
+    if_chain! {
+        if match_trait_method(cx, expr, &paths::ITERATOR);
+
+        if flat_map_args.len() == 2;
+        if let hir::ExprKind::Closure(_, _, body_id, _, _) = flat_map_args[1].node;
```

Comment:
You could insert the `then` of the `if_chain!` here and then split up the two if chains for `|x| x` and `convert::identity`, like this:
```suggestion
        then {
            let arg = flat_map_args[1];
            if_chain! {
                if let hir::ExprKind::Closure(_, _, body_id, _, _) = arg.node;
                // check for `|x| x`
            }
            
            if_chain! {
                if let hir::ExprKind::Path(ref qpath) = arg.node;
                // check for `convert::identity`
            }
        }
```